doc/compiler/control/CheckpointRestoreCoordination.md

```diff
@@ -0,0 +1,434 @@
+Exception [1] and GNU General Public License, version 2 with the
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] https://openjdk.org/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+-->
+
+# Background
+
+This documention outlines how the compiler prepares for checkpoint and restore.
+This is a very dynamic process that requires the coordination of several
+components, as well as taking into account changes in the environment. This
+coordination is important for a few reasons:
+
+* The options can change on restore, and so the compiler threads cannot be in
+flight while the new options are being processed and applied.
+* The CPU restrictions can get lifted on restore, which can result low level CPU
+queries returning different answers within the same compilation.
+* The SCC may no longer exist on restore, or conversely, an SCC that did not
+exist pre-checkpoint now does exist.
+
+For all these reasons, and more, the compiler needs to ensure coordination to
+maintain coherency across a checkpoint/restore.
+
+# High Level Overview
+
+At bootstrap, the compiler registers callbacks for the
+`J9HOOK_VM_PREPARING_FOR_CHECKPOINT` and `J9HOOK_VM_PREPARING_FOR_RESTORE`
+hooks, namely `jitHookPrepareCheckpoint` and `jitHookPrepareRestore`
+respectively.
+
+When the application signals to the VM that a checkpoint should occur, the VM
+calls `jitHookPrepareCheckpoint` as part of the checkpoint process. This calls
+the top level method in the compiler that performs the necessary coordination:
+`TR::CompilationInfo::prepareForCheckpoint`.
+Once the coordination is complete, the executing thread returns to the VM, which
+eventually invokes CRIU to checkpoint itself.
+
+On restore, the VM calls `jitHookPrepareRestore` as part of the restore process.
+This first checks if further checkpointing is not allowed in order to remove the
+portability restrictions on the target CPU. It then calls the top level method
+in the compiler that performs the necessary coordination:
+`TR::CompilationInfo::prepareForRestore()`. Once the coordination is complete,
+the executing thread returns to the VM, which eventually resumes the
+application.
+
+# Preparing for Checkpoint
+
+As mentioned above, `TR::CompilationInfo::prepareForCheckpoint` handles all the
+coordination needed to prepare the compiler for checkpoint. This is a very
+complex set of operations as it involves coordinating multiple threads, as well
+as being aware of events such as shutdown.
+
+First, this method releases VM Access and acquires the Compilation Monitor. It
+then checks if the checkpoint should be interrupted. It uses the
+`TR::CompilationInfo::_checkpointStatus` member variable to determine the state
+of the checkpoint. The Compilation Monitor is used to guard accesses to the
```

Comment:
The next few paragraphs that describe the checkpoint preparation process don't elaborate on how `TR::CompilationInfo::_checkpointStatus` is used to determine the state of the checkpoint. I have a similar comment below that mentions adding different `TR_CheckpointStatus` state in the corresponding paragraphs.

Reply:
I'll put a list of the transitions the state goes through.